### PR TITLE
fix(contract): check address params in initialize

### PR DESCRIPTION
**Pull request by PierrickGT**
Issue: https://github.com/code-423n4/2021-06-pooltogether-findings/issues/81

### CODE REVIEW

**Comment by aodhgan** [comment]
Might be a good idea to note the limitation that `isContract()` does not work when the contract is being constructed

**Comment by PierrickGT** [comment]
> Might be a good idea to note the limitation that `isContract()` does not work when the contract is being constructed

Done: https://github.com/pooltogether/aave-yield-source/pull/19/files#diff-000769aba9936f029eb581d424eae1c5122d9c5a4a28353445d678dcd74aba24R75

File: contracts/yield-source/ATokenYieldSource.sol [hunk 1]
```diff
@@ -93,14 +96,17 @@ contract ATokenYieldSource is ERC20Upgradeable, IProtocolYieldSource, AssetManag
...
+
+    require(address(_lendingPoolAddressesProviderRegistry).isContract(), "ATokenYieldSource/lendingPoolRegistry-not-contract-address");
     lendingPoolAddressesProviderRegistry = _lendingPoolAddressesProviderRegistry;
 
     __Ownable_init();
     transferOwnership(_owner);
 
     __ERC20_init(_name,_symbol);
-    require(_decimals > 0, "ATokenYieldSource/decimals-gt-zero");
+    require(_decimals > 0, "ATokenYieldSource/decimals-not-greater-than-zero");
```
Comment (asselstine) [inline contracts/yield-source/ATokenYieldSource.sol]: keep it short- use just `gt`.  Every byte counts!

File: contracts/yield-source/ATokenYieldSource.sol [hunk 2]
```diff
@@ -189,7 +195,7 @@ contract ATokenYieldSource is ERC20Upgradeable, IProtocolYieldSource, AssetManag
   function supplyTokenTo(uint256 mintAmount, address to) external override nonReentrant {
     uint256 shares = _tokenToShares(mintAmount);
 
-    require(shares > 0, "ATokenYieldSource/shares-equal-zero");
+    require(shares > 0, "ATokenYieldSource/shares-not-greater-than-zero");
```
Comment (asselstine) [inline contracts/yield-source/ATokenYieldSource.sol]: same- `gt`

File: contracts/yield-source/ATokenYieldSource.sol [hunk 3]
```diff
@@ -93,14 +96,17 @@ contract ATokenYieldSource is ERC20Upgradeable, IProtocolYieldSource, AssetManag
     initializer
     returns (bool)
   {
+    require(address(_aToken).isContract(), "ATokenYieldSource/aToken-not-contract-address");
     aToken = _aToken;
+
+    require(address(_lendingPoolAddressesProviderRegistry).isContract(), "ATokenYieldSource/lendingPoolRegistry-not-contract-address");
```
Comment (asselstine) [inline contracts/yield-source/ATokenYieldSource.sol]: Let's skip the `isContract()`.  It adds a lot of gas, and barely does more than a non-zero address check.

**Review by asselstine — CHANGES_REQUESTED** [review]

File: contracts/yield-source/ATokenYieldSource.sol (hunk 3 above)
Comment (asselstine) [reply contracts/yield-source/ATokenYieldSource.sol]: (i.e. just a zero check)

Comment (PierrickGT) [reply contracts/yield-source/ATokenYieldSource.sol]: Done: https://github.com/pooltogether/aave-yield-source/pull/19/commits/54cb60e29a201a704151661d746cfa1fa5263fcd

File: contracts/yield-source/ATokenYieldSource.sol (hunk 2 above)
Comment (PierrickGT) [reply contracts/yield-source/ATokenYieldSource.sol]: Done: https://github.com/pooltogether/aave-yield-source/pull/19/commits/54cb60e29a201a704151661d746cfa1fa5263fcd

File: contracts/yield-source/ATokenYieldSource.sol (hunk 1 above)
Comment (PierrickGT) [reply contracts/yield-source/ATokenYieldSource.sol]: Done: https://github.com/pooltogether/aave-yield-source/pull/19/commits/54cb60e29a201a704151661d746cfa1fa5263fcd

File: contracts/yield-source/ATokenYieldSource.sol [hunk 4]
```diff
@@ -9,6 +9,7 @@ import "@openzeppelin/contracts-upgradeable/math/SafeMathUpgradeable.sol";
 import "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
 import "@openzeppelin/contracts-upgradeable/token/ERC20/IERC20Upgradeable.sol";
 import "@openzeppelin/contracts-upgradeable/token/ERC20/SafeERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/AddressUpgradeable.sol";
```
Comment (asselstine) [inline contracts/yield-source/ATokenYieldSource.sol]: ✂️

File: contracts/yield-source/ATokenYieldSource.sol [hunk 5]
```diff
@@ -21,6 +22,7 @@ import "../interfaces/IProtocolYieldSource.sol";
 /// @dev This contract inherits AssetManager which extends OwnableUpgradable
 /// @notice Yield source for a PoolTogether prize pool that generates yield by depositing into Aave V2
 contract ATokenYieldSource is ERC20Upgradeable, IProtocolYieldSource, AssetManager, ReentrancyGuardUpgradeable {
+  using AddressUpgradeable for address;
```
Comment (asselstine) [inline contracts/yield-source/ATokenYieldSource.sol]: ✂️

**Review by asselstine — CHANGES_REQUESTED** [review]

File: contracts/yield-source/ATokenYieldSource.sol (hunk 4 above)
Comment (PierrickGT) [reply contracts/yield-source/ATokenYieldSource.sol]: Done: https://github.com/pooltogether/aave-yield-source/pull/19/commits/7f7a8714b900beb9b3cc46e11b3f19922afcb2f0

File: contracts/yield-source/ATokenYieldSource.sol (hunk 5 above)
Comment (PierrickGT) [reply contracts/yield-source/ATokenYieldSource.sol]: Done: https://github.com/pooltogether/aave-yield-source/pull/19/commits/7f7a8714b900beb9b3cc46e11b3f19922afcb2f0

**Review by asselstine — APPROVED** [review]

**Pull request merged**